src/functional/swing_slide_filter.zig

```diff
@@ -0,0 +1,422 @@
+        current_segment.end_value = compressed_lines_and_index[index + 1];
+        current_segment.end_time = @intFromFloat(compressed_lines_and_index[index + 2] - 1);
+
+        if (current_segment.start_time < current_segment.end_time) {
+            updateSwingLinearFunction(current_segment, &current_linear_function, 0.0);
+            try decompressed_values.append(current_segment.start_value);
+            var current_timestamp: usize = current_segment.start_time + 1;
+            while (current_timestamp < current_segment.end_time) : (current_timestamp += 1) {
+                const y: f64 = evaluateLinearFunctionAtTime(current_linear_function, current_timestamp);
+                try decompressed_values.append(y);
+            }
+            try decompressed_values.append(current_segment.end_value);
+            first_timestamp = current_timestamp + 1;
+        } else {
+            try decompressed_values.append(current_segment.start_value);
+            first_timestamp += 1;
+        }
+    }
+}
+
+/// Converts `value`:usize to f80
+fn usizeToF80(value: usize) f80 {
+    return @as(f80, @floatFromInt(value));
+}
+
+/// Computes the intercept coefficient of a line that passes through the point
+/// (time, value) and with the given slope coefficient.
+fn computeInterceptCoefficient(slope: f80, time: usize, value: f64) f80 {
+    return value - slope * usizeToF80(time);
+}
+
+/// Updates the linear function coeficients in `line` that passes throught the two points
+/// of the `segment`. The line is swinged down or up based on the `error_bound`.
+/// If `error_bound` is negative, `line` is swing down. It is swing up otherwise.
+fn updateSwingLinearFunction(
+    segment: Segment,
+    linear_function: *LinearFunction,
+    error_bound: f32,
+) void {
+    if (segment.end_time != segment.start_time) {
+        const duration: f80 = @floatFromInt(segment.end_time - segment.start_time);
+        linear_function.slope = (segment.end_value + error_bound - segment.start_value) / duration;
+        linear_function.intercept = segment.start_value - linear_function.slope * usizeToF80(segment.start_time);
+    } else {
+        linear_function.slope = 0.0;
+        linear_function.intercept = segment.start_value;
+    }
+}
+
+/// Evaluate the linear function `line` at `time`.
+fn evaluateLinearFunctionAtTime(linear_function: LinearFunction, time: usize) f64 {
+    return @floatCast(linear_function.slope * usizeToF80(time) + linear_function.intercept);
+}
+
+/// Append `value` of `type` determined at compile time to `compressed_values`.
+fn appendValue(comptime T: type, value: T, compressed_values: *std.ArrayList(u8)) !void {
+    // Compile-time type check
+    switch (@TypeOf(value)) {
+        f64 => {
+            const value_as_bytes: [8]u8 = @bitCast(value);
+            try compressed_values.appendSlice(value_as_bytes[0..]);
+        },
+        usize => {
+            const value_as_bytes: [8]u8 = @bitCast(@as(f64, @floatFromInt(value)));
```

Comment:
Unclear why `index` is converted to a `f64` before it is written to `compressed_values`.

Reply:
I was reading this https://github.com/ziglang/zig/issues/5185 to understand how is the `usize` defined. Could not find the answer clearly. But, in `src/functional/poor_mans_compression.zig`  the index is not cast to f64 and it is assumed that the usize is stored as a 64-bit integer? I will remove the cast and insert the value directly.

Reply:
`usize` is [equivalent](https://ziglang.org/documentation/master/#toc-Primitive-Types) to a `size_t` in [C](https://en.cppreference.com/w/c/types/size_t), so on x86_64 it should be 64 bit. So removing the cast should still work but should helpfully fail to compile if `usize` is another size since it cannot become `[8]u8`.